spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala

```diff
@@ -512,7 +512,22 @@ trait OptimisticTransactionImpl extends TransactionalWrite
 
       val newProtocolForLatestMetadata =
         Protocol(readerVersionAsTableProp, writerVersionAsTableProp)
-      val proposedNewProtocol = protocolBeforeUpdate.merge(newProtocolForLatestMetadata)
+
+      // When the protocol versions are explicitly set on table features protocol we may
+      // downgrade to legacy protocol versions. Legacy protocol versions can only be
+      // used if a table supports *exactly* the set of features in that legacy protocol
+      // version, with no "gaps". By merging in the protocol features from a particular
+      // protocol version, we may end up with such a "gap-free" protocol. E.g. if a table
+      // has only table feature "checkConstraints" (added by writer protocol version 3)
+      // but not "invariants" and "appendOnly", then setting the minWriterVersion to
+      // 2 or 3 will add "invariants" and "appendOnly", filling in the gaps for writer
+      // protocol version 3, and then we can downgrade to version 3.
+      val mergedProtocol = protocolBeforeUpdate.merge(newProtocolForLatestMetadata)
+      val proposedNewProtocol = if (newProtocolForLatestMetadata.supportsWriterFeatures) {
+        mergedProtocol
+      } else {
+        mergedProtocol.downgradeProtocolToLegacyVersionsIfNeeded
```

Comment:
Can we just do this unconditionally? I.e., whenever the protocol happens to be _perfectly_ representable by a legacy version, use the legacy protocol version? And if we have https://github.com/delta-io/delta/pull/2841/files, can this just use `downgradeProtocolVersionsIfNeeded`?

I think the goal should be simply _always use the "most legacy" protocol representation that can represent the feature set_. I don't think we'd call it an _invariant_ (because other implementations may not do this), but we have to try to do this normalization whenever we change the protocol.

Reply:
> Can we just do this unconditionally? I.e., whenever the protocol happens to be perfectly representable by a legacy version, use the legacy protocol version?

This is how I originally did this. The issue was is it changes current behaviour. The result is that setting table features protocol versions to a legacy protocol won't "stick" when there are no table features. For example, currently, setting `(1, 7)` to a table with `(2, 2)` results to `Protocol(2, 7, {"appendOnly", "invariants"})`. With the proposed change it will go back to `(2, 2)` and therefore be a noop. On the other hand, I do not see why someone would want to convert a protocol to table features without adding a particular table feature. It just might be a bit confusing his operation to be noop. So, I ended up removing the condition.

> And if we have https://github.com/delta-io/delta/pull/2841/files, can this just use downgradeProtocolVersionsIfNeeded?

Moving the normalization to `downgradeProtocolVersionsIfNeeded` will result to downgrading protocol versions with "gaps" directly at DROP command without the extra version setting step. That would be a different user journey from the one proposed in the PR. However, I like the concept of normalization. I refactored `downgradeProtocolToLegacyVersionsIfNeeded` to clearly show there is a normalization process when we set protocol versions. Also, it is now clear this does not happen when we call the DROP FEATURE command.